### DISCUSSION TITLE
using "useCallback" with form methods.

### DISCUSSION THREAD

**Discussion by hisothreed** [react-hook-form/react-hook-form#7363] [Q&A]
Hello, i've noticed recently that changing in the form values is causing a re-render to all the components that use the functions provided by the hook, such as "handleSubmit" or "handleChange" or "handleBlur" etc.., i've solved the issue by using useCallback, my question is, is there any downside of using useCallback hook with these methods? and why it wasn't implemented already internally in the library to avoid causing such a very sensitive issue, as in react-native it's really obvious when you have a performance issue that's the main reason that brought me to this library, i've tried useFormik and they don't re-initialize the events thus not having this issue.

Thank you.

**Comment by bluebill1049** [comment]
could you share an example that could potentially lead to a perf issue (codesandbox)? I guess you are refer to `onChange` from Controller or useController?

**Reply by hisothreed** [reply]
this is an example for both cases, with and without `useCallback`, https://codesandbox.io/s/frosty-sun-oqbwy?file=/src/App.js 
found these, also related to this issue: 
https://github.com/react-hook-form/react-hook-form/pull/5113
https://github.com/react-hook-form/react-hook-form/issues/6435

**Reply by bluebill1049** [reply]
![Screen Shot 2021-12-26 at 10 11 04 am](https://user-images.githubusercontent.com/10513364/147395226-dfbbc284-028b-4573-a351-247a6c0398e3.png)

what's the issue presented here? 

as you can see in the original PR we had a lengthy discussion there as well. react hook form isolates re-render at the component/hook level.

**Reply by hisothreed** [reply]
@bluebill1049 sorry, the example was not up to date can you please check it now.

**Reply by hisothreed** [reply]
@bluebill1049 I just want to know if the re-initialization of the callbacks is intended behavior, and using `useCallback` to handle this case to avoid re-rendering expensive components is also ok and there is no downside of memoizing hook or control methods or callbacks, and if there is any reason that explains why this is not taking place internally already to avoid this case?

**Reply by bluebill1049** [reply]
`handleSubmit` was meant to use at the root of your form Component, and hence `useCallback` it's not necessary. However, pass down `handleSubmit` as a prop seems cate for certain usage as what you have demonstrated above. I will send a PR to attach `useCallback` within the lib for `handleSubmit`.

**Reply by bluebill1049** [reply]
adding more context, because `handleSubmit` return callback (a curry function), which is the other reason why it's not memorized, i will see what i can do worse cases scenario it's fine to leave that `useCallback` at your app level.

**Reply by bluebill1049** [reply]
just another approach, why not just pass down `handleSubmit` as a function.

`<Counter handleSubmit={handleSubmit} name={"second"} />` then you can optional memorise your callback function.

**Reply by hisothreed** [reply]
@bluebill1049 also if you notice the updates I made to the example, even the callbacks of other fields are being changed when changing any of the fields, so if I have field1 and field2, changed something in field1, field2 callbacks + ref all are being changed causing it to re-render, all the connected components, so it's not only in handleSubmit, but in all the methods exposed by any of the hooks.

**Reply by hisothreed** [reply]
@bluebill1049 found something actually, using ```register``` instead of ```useController``` will actually work just fine, it will render only the registered component if it was changed, other components won't re-render and it won't re-initialize any of the methods, I've also updated the example to cover this case.


Update: using ```Controller``` is behaving correctly also, so the issue is only with ```useController```, updated the example again.

**Reply by bluebill1049** [reply]
Thanks for the updated can, useController should live at the component level not root of your form or app.

**Reply by hisothreed** [reply]
@bluebill1049 yep totally understand, but still even if I'm using it in a nested component or anything other than the root, I'm expecting that it would not affect other controls (not causing the main issue of re-initializing all the other registered components when they are all stale)

**Reply by bluebill1049** [reply]
<img width="763" alt="Screen Shot 2021-12-27 at 11 13 26 pm" src="https://user-images.githubusercontent.com/10513364/147470812-89325bab-6002-4bb3-86c2-998def62d036.png">

can you provide some more detail or example as what you described above? in your example both `useController` is appearing at the root of your application.

**Reply by hisothreed** [reply]
@bluebill1049 you're right, that's a mis-use from my side apparently using the `useController` in a sub component is not causing any issues with the other controls on higher level, onyl the useController at top level is casuing all the issues, listed all the cases in the example (updated) if you want to experiment yourself, apologies and thank you 🙏🏻

**Reply by bluebill1049** [reply]
no worries at all @husseinAbdulnabi thanks for providing those examples. 👍

**Reply by hisothreed** [reply]
@bluebill1049 very sorry, one last thing, do we have any alternative to `useWatch ` or `watch` to listen for value changes for a specific field? both are triggering re-render even if I provided the name of the field I'm watching for.

my case basically is: I have a field that will show up based on another field value, so a straightforward solution is to use `const value = watch({ name: 'field' })` and then `{!!value && <Controlelr ... /> }` but it's causing a re-render to everything connected on changing any value in the form (also updated the example). so what are my options/alternatives here?

**Reply by bluebill1049** [reply]
follow the doc: https://react-hook-form.com/api/useform/watch there is a video tutorial as well

```
watch((data, { name, type }) => console.log(data, name, type))
```
